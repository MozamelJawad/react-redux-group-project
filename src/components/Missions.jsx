import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions, joinMission } from '../redux/missions/missionsSlice';
import '../redux/missions/missions.css';

function Missions() {
  const { missions, isLoading, error } = useSelector((state) => state.missions);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchMissions());
  }, [dispatch]);

  if (isLoading) {
    return (
      <div className="loading">
        Loading...
      </div>
    );
  }

  if (error) {
    return (
      <div className="error">
        Error:
        {error}
      </div>
    );
  }

  return (
    <div className="table-container">
      <table className="mission-table">
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th className="status">Status</th>
            <th className="status" aria-label="Status">Status</th>
          </tr>
        </thead>
        <tbody>
          {missions.map((mission) => {
            const {
            /* eslint-disable camelcase */
              mission_id, mission_name, description, join,
            } = mission;

            return (
              /* eslint-disable camelcase */
              <tr key={mission_id}>
                <td className="missionName">{mission_name}</td>
                <td className="missionDesc">{description}</td>
                <td className="status">
                  <button
                    type="button"
                    className={`${join ? 'activeMember' : 'notMember'} memberBtn`}
                    aria-label={join ? 'Active Member' : 'Not a Member'}
                  >
                    {join ? 'Active Member' : 'Not a Member'}
                  </button>
                </td>
                <td className="status">
                  <button
                    type="button"
                    className={`leaveBtnMission ${join ? 'joined' : 'notJoined'}`}
                    aria-label={join ? 'Leave Mission' : 'Join Mission'}
                    onClick={() => dispatch(joinMission({ mission_id }))}
                  >
                    {join ? 'Leave Mission' : 'Join Mission'}
                  </button>
                </td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </div>
  );
}

export default Missions;
