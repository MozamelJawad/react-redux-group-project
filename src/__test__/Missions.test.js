import { configureStore } from '@reduxjs/toolkit';
import { render, fireEvent } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import missionsReducer from '../redux/missions/missionsSlice';
import Missions from '../components/Missions';

describe('Render Missions component correctly', () => {
  test('Render missions component with actions correctly', async () => {
    const store = configureStore({
      reducer: {
        missions: missionsReducer,
      },
      initialState: {
        missions: {
          missions: [
            {
              mission_id: '1',
              mission_name: 'Mission A',
              description: 'Mission A Description',
              join: true,
            },
            {
              mission_id: '2',
              mission_name: 'Mission B',
              description: 'Mission B Description',
              join: false,
            },
          ],
          isLoading: false,
          error: '',
        },
      },
    });

    const { container, getByText } = render(
      <Provider store={store}>
        <Router>
          <Missions />
        </Router>
      </Provider>,
    );

    expect(container).toMatchSnapshot();

    // Test join mission button
    const joinMissionButton = getByText('Join Mission');
    fireEvent.click(joinMissionButton);

    // The button text should change to "Leave Mission" after clicking
    expect(getByText('Leave Mission')).toBeInTheDocument();

    // Test leave mission button
    const leaveMissionButton = getByText('Leave Mission');
    fireEvent.click(leaveMissionButton);

    // The button text should change back to "Join Mission" after clicking again
    expect(getByText('Join Mission')).toBeInTheDocument();
  });
});
