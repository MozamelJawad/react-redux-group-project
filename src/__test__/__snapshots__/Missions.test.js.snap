// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Missions component correctly Render missions component with actions correctly 1`] = `
<div>
  <div
    class="table-container"
  >
    <table
      class="mission-table"
    >
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th
            class="status"
          >
            Status
          </th>
          <th
            aria-label="Status"
            class="status"
          >
            Status
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
